File: Regex/Parser.cpp
#include "Parser.hpp"

#include <sstream>
#include <iostream>
#include <algorithm>

namespace Regex {

    std::unique_ptr<Parser::Node> Parser::parseSymbol(const std::string &regex, int &pos)
    {
        if(pos >= regex.size()) {
            throw ParseException("Expected symbol", pos);
        }

        if(regex[pos] == ')' || regex[pos] == '|' || regex[pos] == '*' || regex[pos] == '+' || regex[pos] == '?') {
            std::stringstream ss;
            ss << "Unexpected character '" << regex[pos] << "'";
            throw ParseException(ss.str(), pos);
        }

        if(regex[pos] == '[') {
            return parseCharacterClass(regex, pos);
        } else if(regex[pos] == '\\') {
            return parseEscape(regex, pos);
        } else {
            Symbol symbol = regex[pos];
            pos++;

            return std::make_unique<SymbolNode>(symbol);
        }
    }

    std::vector<Parser::CharacterClassNode::Range> invertRanges(std::vector<Parser::CharacterClassNode::Range> &ranges)
    {
        std::vector<Parser::CharacterClassNode::Range> outputRanges;
        std::sort(ranges.begin(), ranges.end(), [](const Parser::CharacterClassNode::Range &a, const Parser::CharacterClassNode::Range &b) { return a.first < b.first; });
        Parser::Symbol start = 0;
        for(const auto &range : ranges) {
            if(range.first > start) {
                outputRanges.push_back(Parser::CharacterClassNode::Range(start, range.first - 1));
            }
            start = range.second + 1;
        }
        outputRanges.push_back(Parser::CharacterClassNode::Range(start, 127));
        return outputRanges;
    }

    std::unique_ptr<Parser::Node> Parser::parseCharacterClass(const std::string &regex, int &pos)
    {
        if(regex[pos] != '[') {
            throw ParseException("Expected [", pos);
        }

        pos++;

        if(pos >= regex.size()) {
            throw ParseException("Expected ]", pos);
        }

        bool invert = false;
        if(regex[pos] == '^') {
            invert = true;
            pos++;
        }

        std::vector<CharacterClassNode::Range> ranges;
        while(true) {
            if(pos >= regex.size()) {
                throw ParseException("Expected ]", pos);
            }
            if(regex[pos] == ']') {
                pos++;
                break;
            }

            Symbol start = regex[pos];
            Symbol end = start;

            pos++;
            if(pos < regex.size() && regex[pos] == '-') {
                pos++;
                if(pos >= regex.size()) {
                    throw ParseException("Expected symbol", pos);
                }

                end = regex[pos];
                pos++;
            }
            ranges.push_back(CharacterClassNode::Range(start, end));
        }

        if(invert) {
            ranges = invertRanges(ranges);
        }

        return std::make_unique<CharacterClassNode>(std::move(ranges));
    }

    std::unique_ptr<Parser::Node> Parser::parseEscape(const std::string &regex, int &pos)
    {
        if(regex[pos] != '\\') {
            throw ParseException("Expected \\", pos);
        }

        pos++;

        if(pos >= regex.size()) {
            throw ParseException("Incomplete escape", pos);
        }

        std::vector<CharacterClassNode::Range> ranges;
        bool invert = false;
        switch(regex[pos]) {
            case 'S': invert = true;
            case 's':
                ranges.push_back(CharacterClassNode::Range(' ', ' '));
                ranges.push_back(CharacterClassNode::Range('\t', '\t'));
                break;

            case 'W': invert = true;
            case 'w':
                ranges.push_back(CharacterClassNode::Range('a', 'z'));
                ranges.push_back(CharacterClassNode::Range('A', 'Z'));
                ranges.push_back(CharacterClassNode::Range('0', '9'));
                ranges.push_back(CharacterClassNode::Range('_', '_'));
                break;
        }
        if(ranges.size() > 0) {
            pos++;
            if(invert) {
                ranges = invertRanges(ranges);
            }
            return std::make_unique<CharacterClassNode>(std::move(ranges));
        }

        Symbol symbol;
        switch(regex[pos]) {
            case 't': symbol = '\t'; break;
            case 'n': symbol = '\n'; break;
            case 'r': symbol = '\r'; break;
            default: symbol = regex[pos]; break;
        }
        pos++;

        return std::make_unique<SymbolNode>(symbol);        
    }

    std::unique_ptr<Parser::Node> Parser::parseOneOf(const std::string &regex, int &pos)
    {
        if(regex[pos] == '(') {
            std::vector<std::unique_ptr<Node>> nodes;
            pos++;
            while(true) {
                nodes.push_back(parseSequence(regex, pos));
                
                if(pos < regex.size()) { 
                    if(regex[pos] == ')') {
                        pos++;
                        break;
                    }
                    if(regex[pos] == '|') {
                        pos++;
                        continue;
                    }
                }
                
                throw ParseException("Expected | or )", pos);
            }

            if(nodes.size() == 1) {
                return std::move(nodes[0]);
            } else {
                return std::make_unique<OneOfNode>(std::move(nodes));
            }
        } else {
            return parseSymbol(regex, pos);
        }
    }

    std::unique_ptr<Parser::Node> Parser::parseSuffix(const std::string &regex, int &pos)
    {
        std::unique_ptr<Node> node = parseOneOf(regex, pos);
        while(pos < regex.size()) {
            if(regex[pos] == '*') {
                node = std::make_unique<ZeroOrMoreNode>(std::move(node));
                pos++;
            } else if(regex[pos] == '+') {
                node = std::make_unique<OneOrMoreNode>(std::move(node));
                pos++;
            } else if(regex[pos] == '?') {
                node = std::make_unique<ZeroOrOneNode>(std::move(node));
                pos++;
            } else {
                break;
            }
        }

        return node;
    }

    std::unique_ptr<Parser::Node> Parser::parseSequence(const std::string &regex, int &pos)
    {
        std::vector<std::unique_ptr<Node>> nodes;
        while(pos < regex.size() && regex[pos] != '|' && regex[pos] != ')') {
            nodes.push_back(parseSuffix(regex, pos));
        }

        if(nodes.size() == 1) {
            return std::move(nodes[0]);
        } else {
            return std::make_unique<SequenceNode>(std::move(nodes));
        }
    }

    std::unique_ptr<Parser::Node> Parser::parse(const std::string &regex)
    {
        int pos = 0;
        std::unique_ptr<Node> node = parseSequence(regex, pos);
        if(pos != regex.size()) {
            std::stringstream ss;
            ss << "Unexpected character '" << regex[pos] << "'";
            throw ParseException(ss.str(), pos);
        }

        return node;
    }

    void Parser::Node::print(int depth) const
    {
        for(int i=0; i<depth; i++) std::cout << "  ";

        switch(type) {
            case Node::Type::Symbol:
                std::cout << "Symbol: " << static_cast<const SymbolNode*>(this)->symbol << std::endl;
                break;

            case Node::Type::CharacterClass:
                std::cout << "CharacterClass: ";

                for(const auto &r : static_cast<const CharacterClassNode*>(this)->ranges) {
                    Symbol start = r.first;
                    Symbol end = r.second;
                    if(start == end) {
                        std::cout << start << " ";
                    } else {
                        std::cout << start << "-" << end << " ";
                    }
                }
                std::cout << std::endl;
                break;
    
            case Node::Type::Sequence:
                std::cout << "Sequence:" << std::endl;
                for(const auto &n : static_cast<const SequenceNode*>(this)->nodes) {
                    n->print(depth + 1);
                }
                break;
            
            case Node::Type::ZeroOrOne:
                std::cout << "ZeroOrOne:" << std::endl;
                static_cast<const ZeroOrOneNode*>(this)->node->print(depth + 1);
                break;

            case Node::Type::ZeroOrMore:
                std::cout << "ZeroOrMore:" << std::endl;
                static_cast<const ZeroOrMoreNode*>(this)->node->print(depth + 1);
                break;

            case Node::Type::OneOrMore:
                std::cout << "OneOrMore:" << std::endl;
                static_cast<const OneOrMoreNode*>(this)->node->print(depth + 1);
                break;

            case Node::Type::OneOf:
                std::cout << "OneOf:" << std::endl;
                for(const auto &n : static_cast<const OneOfNode*>(this)->nodes) {
                    n->print(depth + 1);
                }
                break;
        }
    }
}

File: src/regex/parser.rs
pub struct Parser;

type Symbol = char;
type Range = (Symbol, Symbol);

pub enum Node {
    Symbol(Symbol),
    CharacterClass(Vec<Range>),
    Sequence(Vec<Node>),
    ZeroOrOne(Box<Node>),
    ZeroOrMore(Box<Node>),
    OneOrMore(Box<Node>),
    OneOf(Vec<Node>)
}

pub struct ParseError {
    message: String,
    pos: usize
}

impl ParseError {
    pub fn new(message: &str, pos: usize) -> ParseError {
        ParseError { message: message.to_string(), pos }
    }
}

impl Parser {
    pub fn parse(regex: &str) -> Result<Node, ParseError> {
        let mut pos = 0;
        let node = Self::parse_sequence(regex, &mut pos)?;
        if pos == regex.len() {
            return Ok(node);
        } else {
            let c = regex.chars().nth(pos).unwrap();
            return Err(ParseError::new(format!("Unexpected character {c}").as_str(), pos));
        }
    }

    fn parse_sequence(regex: &str, pos: &mut usize) -> Result<Node, ParseError> {
        let mut nodes = Vec::new();
        while *pos < regex.len() {
            match regex.chars().nth(*pos).unwrap() {
                '|' | ')' => break,
                _ => nodes.push(Self::parse_suffix(regex, pos)?)
            }
        }
        
        if nodes.len() == 1 {
            return Ok(nodes.pop().unwrap());
        } else {
            return Ok(Node::Sequence(nodes));
        }
    }

    fn parse_suffix(regex: &str, pos: &mut usize) -> Result<Node, ParseError> {
        let mut node = Self::parse_one_of(regex, pos)?;
        while *pos < regex.len() {
            let c = regex.chars().nth(*pos).unwrap();
            match c {
                '*' => node = Node::ZeroOrMore(Box::new(node)),
                '+' => node = Node::OneOrMore(Box::new(node)),
                '?' => node = Node::ZeroOrOne(Box::new(node)),
                _ => break
            }
            *pos += 1;
        }

        return Ok(node);
    }

    fn parse_one_of(regex: &str, pos: &mut usize) -> Result<Node, ParseError> {
        if regex.chars().nth(*pos).unwrap() == '(' {
            let mut nodes = Vec::new();
            *pos += 1;
            loop {
                nodes.push(Self::parse_sequence(regex, pos)?);
                match regex.chars().nth(*pos).unwrap_or_default() {
                    '|' => *pos += 1,
                    ')' => { *pos += 1; break; },
                    _ => return Err(ParseError::new("Expected | or )", *pos))
                }
            }

            if nodes.len() == 1 {
                return Ok(nodes.pop().unwrap());
            } else {
                return Ok(Node::OneOf(nodes));
            }
        } else {
            return Self::parse_symbol(regex, pos);
        }
    }

    fn parse_symbol(regex: &str, pos: &mut usize) -> Result<Node, ParseError> {
        if *pos >= regex.len() {
            return Err(ParseError::new("Expected symbol", *pos));
        }
        
        let c = regex.chars().nth(*pos).unwrap();
        match c {
            ')' | '|' | '*' | '+' | '?' => return Err(ParseError::new(format!("Unexpected character {c}").as_str(), *pos)),
            '[' => return Self::parse_character_class(regex, pos),
            '\\' => return Self::parse_escape(regex, pos),
            _ => {
                let symbol = c as Symbol;
                *pos += 1;

                return Ok(Node::Symbol(symbol));
            }
        }
    }

    fn parse_character_class(regex: &str, pos: &mut usize) -> Result<Node, ParseError> {
        if regex.chars().nth(*pos).unwrap() != '[' {
            return Err(ParseError::new("Expected [", *pos));
        }
        *pos += 1;
        
        if *pos >= regex.len() {
            return Err(ParseError::new("Expected ]", *pos));
        }

        let invert;
        if regex.chars().nth(*pos).unwrap() == '^' {
            invert = true;
            *pos += 1;
        } else {
            invert = false;
        }

        let mut ranges = Vec::new();
        loop {
            if *pos >= regex.len() {
                return Err(ParseError::new("Expected ]", *pos));
            }

            if regex.chars().nth(*pos).unwrap() == ']' {
                *pos += 1;
                break;
            }

            let start = regex.chars().nth(*pos).unwrap() as Symbol;
            let end;
            *pos += 1;
            
            if regex.chars().nth(*pos).unwrap_or_default() == '-' {
                *pos += 1;
                if *pos >= regex.len() {
                    return Err(ParseError::new("Expected symbol", *pos));
                }
                end = regex.chars().nth(*pos).unwrap() as Symbol;
                *pos += 1;
            } else {
                end = start;
            }

            ranges.push((start, end) as Range);
        }

        if invert {
            ranges = Self::invert_ranges(ranges);
        }

        return Ok(Node::CharacterClass(ranges));
    }

    fn invert_ranges(mut ranges: Vec<Range>) -> Vec<Range> {
        let mut output_ranges = Vec::new();
        ranges.sort_by_key(|a| a.0);
        let mut start = 0 as Symbol;
        for (range_start, range_end) in ranges {
            if range_start > start {
                output_ranges.push((start, (range_start as u8 - 1) as Symbol));
            }
            start = (range_end as u8 + 1) as Symbol;
        }
        output_ranges.push((start, 127 as Symbol));
        return output_ranges;
    }

    fn parse_escape(regex: &str, pos: &mut usize) -> Result<Node, ParseError> {
        if regex.chars().nth(*pos).unwrap() != '\\' {
            return Err(ParseError::new("Expected \\", *pos));
        }
        *pos += 1;

        if *pos >= regex.len() {
            return Err(ParseError::new("Incomplete escape", *pos));
        }
        
        let mut ranges = Vec::new();
        let mut invert = false;

        match regex.chars().nth(*pos).unwrap() {
            'S' => { invert = true; ranges.extend_from_slice(&[('\n', '\n'), (' ', ' '), ('\t', '\t')]); },
            's' => { ranges.extend_from_slice(&[(' ', ' '), ('\t', '\t')]); },
            'W' => { invert = true; ranges.extend_from_slice(&[('a', 'z'), ('A', 'Z'), ('0', '9'), ('_', '_')]); },
            'w' => { ranges.extend_from_slice(&[('a', 'z'), ('A', 'Z'), ('0', '9'), ('_', '_')]); },
            _ => ()
        }
        
        if ranges.len() > 0 {
            *pos += 1;
            if invert {
                ranges = Self::invert_ranges(ranges);
            }
            return Ok(Node::CharacterClass(ranges));
        }

        let c = regex.chars().nth(*pos).unwrap();
        let symbol = match c {
            't' => '\t',
            'n' => '\n',
            'r' => '\r',
            _ => c
        };
        *pos += 1;

        return Ok(Node::Symbol(symbol));
    }
}